fuzz/fuzz_targets/ruff_parse_idempotency.rs

```diff
@@ -0,0 +1,46 @@
+//! Fuzzer harness which searches for situations where the parser does not parse or unparse a
+//! particular source snippet consistently.
+
+#![no_main]
+
+use libfuzzer_sys::{fuzz_target, Corpus};
+use ruff_python_ast::source_code::round_trip;
+
+fn do_fuzz(case: &[u8]) -> Corpus {
+    let Ok(code) = std::str::from_utf8(case) else { return Corpus::Reject; };
+
+    // round trip it once to get a formatted version
+    if let Ok(first) = round_trip(code, "fuzzed-source.py") {
+        // round trip it a second time to get a case to compare against
+        if let Ok(second) = round_trip(&first, "fuzzed-source.py") {
+            if cfg!(feature = "full-idempotency") {
+                // potentially, we don't want to test for full idempotency, but just for unsteady states
+                // enable the "full-idempotency" feature when fuzzing for full idempotency
+                assert_eq!(
+                    first, second,
+                    "\nbefore: {:?}\nfirst: {:?}\nsecond: {:?}",
+                    code, first, second
+                );
+            } else if first != second {
+                // by the third time we've round-tripped it, we shouldn't be introducing any more
+                // changes; if we do, then it's likely that we're in an unsteady parsing state
+                let third = round_trip(&second, "fuzzed-source.py")
+                    .expect("Couldn't round-trip the processed source.");
+                assert_eq!(
+                    second, third,
+                    "\nbefore: {:?}\nfirst: {:?}\nsecond: {:?}\nthird: {:?}",
+                    code, first, second, third
+                );
```

Comment:
Idea: You could use similar to produce a diff

https://github.com/charliermarsh/ruff/blob/466719247bc68c0b608bc85be341c5aa5e2a5ec2/crates/ruff_python_formatter/src/lib.rs#L212-L215